base/sort.jl

```diff
@@ -529,75 +530,107 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
+
+    if lt(o, b, a)
+        b, a = a, b
+        lo, mi = mi, lo
+    end
 
-        if lt(o, v[hi], v[lo])
-            if lt(o, v[hi], v[mi])
-                v[hi], v[lo], v[mi] = v[lo], v[mi], v[hi]
-            else
-                v[hi], v[lo] = v[lo], v[hi]
-            end
-        end
+    # a is stably less than b
 
-        # return the pivot
-        return v[lo]
+    if lt(o, c, a)
+        lo
+    elseif lt(o, c, b)
+        hi
+    else
+        mi
     end
 end
 
-# partition!
+# select a pivot, partition v[lo:hi] according
+# to the pivot, and store the result in t[lo:hi].
 #
-# select a pivot, and partition v according to the pivot
-
-function partition!(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
-    pivot = selectpivot!(v, lo, hi, o)
-    # pivot == v[lo], v[hi] > pivot
-    i, j = lo, hi
-    @inbounds while true
-        i += 1; j -= 1
-        while lt(o, v[i], pivot); i += 1; end;
-        while lt(o, pivot, v[j]); j -= 1; end;
-        i >= j && break
-        v[i], v[j] = v[j], v[i]
+# returns (pivot, pivot_index) where pivot_index is the location the pivot
+# should end up, but does not set t[pivot_index] = pivot
+function partition!(t::AbstractVector, lo::Integer, hi::Integer, o::Ordering, v::AbstractVector, rev::Bool)
+    pivot_index = select_pivot(v, lo, hi, o)
+    trues = 0
+    @inbounds begin
+        pivot = v[pivot_index]
+        while lo < pivot_index
+            x = v[lo]
+            fx = rev ? !lt(o, x, pivot) : lt(o, pivot, x)
+            t[(fx ? hi : lo) - trues] = x
+            trues += fx
+            lo += 1
+        end
+        while lo < hi
+            x = v[lo+1]
+            fx = rev ? lt(o, pivot, x) : !lt(o, x, pivot)
+            t[(fx ? hi : lo) - trues] = x
+            trues += fx
+            lo += 1
+        end
     end
-    v[j], v[lo] = pivot, v[j]
 
-    # v[j] == pivot
-    # v[k] >= pivot for k > j
-    # v[i] <= pivot for i < j
-    return j
+    # pivot_index = lo-trues
+    # t[pivot_index] is whatever it was before
+    # t[<pivot_index] <* pivot, stable
+    # t[>pivot_index] >* pivot, reverse stable
+
+    pivot, lo-trues
 end
 
-function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::QuickSortAlg, o::Ordering)
-    @inbounds while lo < hi
-        hi-lo <= SMALL_THRESHOLD && return sort!(v, lo, hi, SMALL_ALGORITHM, o)
-        j = partition!(v, lo, hi, o)
-        if j-lo < hi-j
-            # recurse on the smaller chunk
-            # this is necessary to preserve O(log(n))
-            # stack space in the worst case (rather than O(n))
-            lo < (j-1) && sort!(v, lo, j-1, a, o)
+# Needed for bootstrapping because `view` is not available.
+function reverse_view!(v, lo, hi)
+    for i in 0:div(hi-lo-1, 2)
+        @inbounds v[lo+i], v[hi-i] = v[hi-i], v[lo+i]
+    end
+    v
+end
+
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort, o::Ordering, t::AbstractVector=similar(v), swap=false, rev=false)
```

Comment:
Would it make sense to change `rev` to `..., rev::Val(S)=Val(false)) where S` to compile two versions and remove some branching from `partition!`?

Reply:
One obstacle to that is that the [`rev = !rev`](https://github.com/LilithHafner/julia/blob/afbed36bfe68d8b66bb977b71fbce19b50b93e22/base/sort.jl#L624) lines would introduce type instability. Nevertheless, I tested making this change and commenting out those (stability tests failed) but I did not measure a performance increase. I imagine the compiler is already doing this or something like it for us. When I added branching based on `rev` into the inner loop of `partition!` I didn't notice any regression, so I doubt the compiler is actually putting a branch there.

Reply:
`rev = !rev` can be solved a local variable, like `rev2 = rev == Val(true)` (known at compilation time), but is not necessary if you don't see any regression.